lib/Conversion/FIRRTLToHW/LowerToHW.cpp

```diff
@@ -1650,42 +1650,79 @@ Value FIRRTLLowering::getReadInOutOp(Value v) {
-                                        std::function<void(void)> body,
-                                        std::function<void(void)> resetBody) {
-  auto &op = alwaysFFBlocks[std::make_tuple(
-      builder.getBlock(), clockEdge, clock, resetStyle, resetEdge, reset)];
-  if (op) {
-    runWithInsertionPointAtEndOfBlock(body, op.bodyBlk());
-    runWithInsertionPointAtEndOfBlock(resetBody, op.resetBlk());
+void FIRRTLLowering::addToAlwaysBlock(sv::EventControl clockEdge, Value clock,
+                                      ::ResetType resetStyle,
+                                      sv::EventControl resetEdge, Value reset,
+                                      std::function<void(void)> body,
+                                      std::function<void(void)> resetBody) {
+  auto &op = alwaysBlocks[std::make_tuple(builder.getBlock(), clockEdge, clock,
+                                          resetStyle, resetEdge, reset)];
+  auto &alwaysOp = op.first;
+  auto &resetBodyIfOp = op.second;
+  if (alwaysOp) {
+    if (reset) {
+      assert(resetBodyIfOp.hasValue() &&
+             "reset body must be initialized before");
+      runWithInsertionPointAtEndOfBlock(resetBody, resetBodyIfOp->thenRegion());
+      runWithInsertionPointAtEndOfBlock(body, resetBodyIfOp->elseRegion());
+    } else {
+      runWithInsertionPointAtEndOfBlock(body, alwaysOp.body());
+    }
 
     // Move the earlier always block(s) down to where the last would have been
     // inserted.  This ensures that any values used by the always blocks are
     // defined ahead of the uses, which leads to better generated Verilog.
-    op->moveBefore(builder.getInsertionBlock(), builder.getInsertionPoint());
+    alwaysOp->moveBefore(builder.getInsertionBlock(),
+                         builder.getInsertionPoint());
   } else {
     if (reset) {
-      op = builder.create<sv::AlwaysFFOp>(clockEdge, clock, resetStyle,
-                                          resetEdge, reset, body, resetBody);
+      assert(resetStyle != ::ResetType::NoReset);
+      // Here, we want to create the folloing op. If reset is async, we need to
+      // add reset to a sensitivity list.
+      //
+      // sv.always @(clockEdge or reset) {
+      //   sv.if (reset) {
+      //     resetBody
+      //   } else {
+      //     body
+      //   }
+      // }
+
+      // It is necessary to construct then and else region regardless of
+      // `restBody` and `body`.
+      auto createResetBody = [&]() {
+        if (resetBody)
+          resetBody();
+      };
+      auto createBody = [&]() {
+        if (body)
+          body();
+      };
+
+      if (resetStyle == ::ResetType::AsyncReset) {
+        sv::EventControl events[] = {clockEdge, resetEdge};
+        Value clocks[] = {clock, reset};
+
+        alwaysOp = builder.create<sv::AlwaysOp>(events, clocks, [&] {
+          // If reset is asnyc and negative-edge, then flip the reset value.
+          if (resetEdge == sv::EventControl::AtNegEdge) {
```

Comment:
I'm not sure we actually handle this case(async reset + negative edge). Is there any example in firrtl?

Reply:
As far as I know, FIRRTL cannot hit this. It will always try to lower to posedge reset.  You may want to change this to use an `llvm_unreachable()` as this is intended to _never_ be seen from FIRRTL.